src/yaml.cr

```diff
@@ -147,4 +156,15 @@ module YAML
   def self.dump(object, io : IO)
     object.to_yaml(io)
   end
+
+  # Checks to see if the value is reserved
+  def self.reserved_value?(value)
+    if v = YAML::RESERVED_VALUES.find { |v| v == value }
+      v
+    elsif (['.', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9'] & value[0..1].chars).first?
```

Comment:
This is even more expensive than before your change. Every single string parsed from YAML has to go through this check. Surely there's a better way than all this. Even a regex could be faster here.

Reply:
Perhaps we should have a stricter check for parsing to make sure we don't produce false alarms but always emit quotes if the first character of the value is alpha.

Reply:
@RX14 I don't think aggressive quotes is the right answer, I like the simplicity of YAML, and I think it is safe to use plain scalars when available. A regex is more performant than an array comparison in this case?

Reply:
Yes, a regex is likely more performant than an array comparison, TWO number parses and a time parse. A custom-written character-by-character check will probably be faster if you can work out a simple way to write the regex.

Reply:
@RX14 Last time I tried to use a regex in this PR, I got ridiculed by @asterite. With that in mind and based on some benchmarks that he ran, I don't feel comfortable adding them back in. Let me sleep on it and I will see if I can come up with a faster solution in the am.

Reply:
Why not this?

```crystal
elsif (c = value[0]? && (c.ascii_number? || c == '.'))
```

What @RX14 probably means is that using array literals is slow because they are created over and over. Using a tuple is a bit better, but here you just want to compare a char value.

Reply:
No I mean that the whole method is expensive. The current implementation iterates the string 4 times just to work out whether it should be quoted or not. That's unacceptable.

Reply:
I think we are caring too much about performance here. Let's make it work first with reasonable performance, then we can optimize it, with benchmarks.

In the past, while developing Crystal, I cared too much about performance. Yes, the Base64 encoder is really fast, and other parts of the language are fast too, but the language itself has many problems and inconsistencies. I wish I had paid less attention to performance and more to correctness...

Reply:
Then why not use regex here? It'll probably be simpler to understand.

Reply:
Sure, go ahead :)

Reply:
I'd personally prefer to have a few overzealous quotes in edge cases than use a regex here. If we decide that we don't need to check `YAML.reserved_value?` on parse, then we *can* use heuristics on checking the first 4 or so characters. I'm not saying quote every time but *always* quoting if we start with numbers seems sensible to me.